Limit num of bytes read in _consume_request.

twosigmajab opened the pull request:
Limit num of bytes read in _consume_request.

Fixes #11.

Also, only consume the request for naive clients that do not send the
Expect: 100-continue header.

Finally, when no Content-Length header is sent (e.g. for chunked
Transfer-Encoding), still try to consume the request (up to the limit).
Fixes #12.

twosigmajab commented on wsgi_kerberos.py:
```diff
@@ -23,7 +23,7 @@ def ensure_bytestring(s):
 
 
-def _consume_request(environ):
+def _consume_request(environ, chunklen=1024, read_max=1024*1024):
```
Looks like golang uses a read max of 2 KiB (weirdly written as `2 << 10`):
https://github.com/golang/go/blob/ee3dded36d69264998c39af0ec851371850d842b/src/net/http/client.go#L616-L625
Can reduce this default to match that in my next round of addressing review feedback.

twosigmajab replied:
Done in the latest revision.

propertone reviewed (changes requested):
This change appears to break unit tests for me (test_forbidden hangs in the receive loop)

twosigmajab commented:
Sorry about that, fixed! Another look?

propertone reviewed (commented):
Since you're limiting the max length of the data being read, do we still need to read the data in chunks? Presumably, you can read up to 2k in one call.

jab commented:
@propertone, the behavior of `sock.read(N)` is to "receive **up to** N bytes from the socket", so you can't necessarily read all 2k in one call. (The reason `read` returns the actual number of bytes read is that sometimes it is not able to read all N bytes you requested (even when N is small). So that's how you're able to tell when additional `read` calls are required.)

We could still get rid of the read loop if you prefer to make even less of an effort to prevent affected clients from seeing "broken pipe". Happy to do that if you prefer, just let me know.

propertone reviewed (approved):
I think this is fine as is - thanks for the contribution! Will publish a new release after merging this PR.

propertone reviewed (changes requested):
Actually, I just realized one more thing - the golang comparison is faulty because that's the client code, not the server-code.
The server-side code actually has a [10MB limit](https://github.com/golang/go/blob/master/src/net/http/request.go#L1204), which seems more reasonable. Maybe the limit can be a top-level constant, so someone could technically change it if necessary (such case would probably be rarely enough that a class variable is probably unnecessary).

twosigmajab commented:
🤦 good catch, thanks @propertone! Fixed the default value in the latest revision. Rather than exposing a mutable, global module variable for the case where a user needs to override the default value, I think exposing an optional constructor param is a safer API, even if it is rarely needed.

propertone commented on wsgi_kerberos.py:
```diff
@@ -44,19 +44,29 @@ def _consume_request(environ):
+        else:
+            content_length = int(content_length)
+            if content_length > read_max:
+                return
```
Is that still the right thing to do? I think other web servers just read all data up to max length. On the other hand, if the client will get a connection reset anyway, might as well do it early.

twosigmajab replied:
> On the other hand, if the client will get a connection reset anyway, might as well do it early.

Exactly. I left a comment to make this clearer.

propertone commented on wsgi_kerberos.py:
```diff
@@ -77,10 +87,15 @@ class KerberosAuthMiddleware(object):
+        read and discard up to this many bytes of the request to keep clients
+        from seeing "broken pipe" errors. The default value is 10 MB, the same
+        value that go uses. Ref:
+        https://github.com/golang/go/blob/bc7e4d9/src/net/http/request.go#L1204
```
The golang ref is TMI - put it as a code comment instead. Also add `:type` for consistency.

twosigmajab replied:
Done in the latest revision.

propertone commented on wsgi_kerberos.py:
```diff
@@ -77,10 +87,15 @@ class KerberosAuthMiddleware(object):
 
     def __init__(self, app, hostname='', unauthorized=None, forbidden=None,
-                 auth_required_callback=None):
+                 auth_required_callback=None, max_bytes_read_on_40x=10 << 20):
```
Trying to think of a slightly more friendly name here...`limit_request_size` ([apache](https://httpd.apache.org/docs/2.4/mod/core.html#limitrequestbody)), `max_body_size` ([nginx](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size)), `max_content_length` ([flask](https://flask.palletsprojects.com/en/1.1.x/config/#MAX_CONTENT_LENGTH))...something more along those lines would be better.

twosigmajab replied:
What do you think of `read_max_on_401`?

I did think carefully about the name for this before I chose it. I think the issue with the three shorter examples you gave is that they don't mean the same thing: they mean an unconditional limit on all requests, not just unauthenticated ones as in this case. I hope `read_max_on_401` is a good compromise between conciseness and preciseness.

propertone reviewed (changes requested):
Few other comments:
* Given the various issues uncovered so far, I'm not confident that this code is correct for various input permutations of body size and content-length. Can you add at least 1 unit test for it?
* One of the issues this PR supposedly addresses is chunked transfer encoding, but have you confirmed that this actually works as intended in practice? I wasn't able to reproduce this since wsgi just reads the entire body.

twosigmajab commented:
@propertone, thinking about this some more, I actually propose a different approach; please see #25. I don't think we should do this in WSGI-Kerberos, since it is already done in production-quality WSGI servers, and doing it in WSGI-Kerberos defeats that. WSGI servers (or even reverse proxies like NGINX) are the proper place to do this; they can actually implement this more safely, since they have raw socket access.

twosigmajab commented on test_wsgi_kerberos.py:
```diff
@@ -120,6 +120,16 @@ def test_unauthorized(self):
+        resp = app.post('/', 'unread request body', expect_errors=True)
+        self.assertEqual(resp.status_int, 401)
+        self.assertGreater(resp.request.body_file.raw.remaining, 0)
+
```
Added a passing test for this.

twosigmajab commented on wsgi_kerberos.py:
```diff
@@ -40,25 +40,39 @@ def _consume_request(environ):
     client implementations.
     '''
+    if environ.get("HTTP_EXPECT") == "100-continue":
+        return
```
I noticed and fixed this additional issue with the current implementation. Just a few lines above, the quote from the spec says this should only be done for a request that does not include an "Expect: 100-continue" header. We should not attempt to consume *Content-Length* bytes of request bodies sent by clients that send "Expect: 100-continue", since they wait for a "100 Continue" response before actually sending the body.

twosigmajab commented on wsgi_kerberos.py:
```diff
@@ -40,25 +40,39 @@ def _consume_request(environ):
             except socket.error as err:
                 if err.errno != errno.EAGAIN:
                     break
-    except (KeyError, ValueError):
```
(KeyError looks like it was mistakenly left in here after e.g. `environ['CONTENT_LENGTH']` was changed to `environ.get('CONTENT_LENGTH')` above in an earlier revision? And ValueError looks like it's trying to account for the `int(content_length_string)` conversion, but I think WSGI servers should already validate that Content-Length headers contain valid integers before we get here?)

It looks like this should just catch Exception in any case -- this is a best effort function that should not ever cause an unhandled exception and resulting 500 error on the server. I made that improvement below, and also added a debug log statement with the exception info so the user actually has a way of finding out more about any swallowed exceptions.

propertone replied:
That's fine, though I'm not sure about printing the stack trace (even in debug mode) - it might confuse developer into thinking there is an actual problem when there isn't. Should be fine to just print just the message. I'd also be explicitly about the fact that the exception comes from wsgi_kerberos so devs don't have to chase that down, eg: `wsgi_kerberos:_consume_request failed: str(exc)`

It might actually be better to just print nothing at all (as before). Unless you're a wsgi_kerberos developer, I can't imagine why devs would care about seeing debug logs of clients sending invalid content-length headers.

propertone replied:
> I think WSGI servers should already validate that Content-Length headers contain valid integers before we get here

I wouldn't rely on that - at least the wsgiref server doesn't do any validation.

twosigmajab reviewed (commented):
I believe the latest revision addresses all the latest review comments. Another look?

twosigmajab commented:
> One of the issues this PR supposedly addresses is chunked transfer encoding, but have you confirmed that this actually works as intended in practice? I wasn't able to reproduce this since wsgi just reads the entire body.

What servers did you test with? Support was added for this to several popular implementations in the past couple years. If you try with e.g. Gunicorn and Flask I think you should be able to reproduce. Please see e.g. https://github.com/benoitc/gunicorn/issues/1653 (which I also linked from https://github.com/deshaw/wsgi-kerberos/issues/12#issuecomment-524045298).

(Of course, it also matters which clients you test with. For example, with curl you need to [explicitly add the "Transfer-Encoding: Chunked" header](https://stackoverflow.com/questions/31504531/is-it-possible-to-upload-a-file-with-curl-from-a-pipe). Also, curl is *not* the type of naive client (that blindly uploads large request bodies and expects them to be consumed before checking if authentication was successful) that we are trying to help here. See e.g. https://gms.tf/when-curl-sends-100-continue.html for more info. I guess WSGI-Kerberos does not currently support sending the 100 Continue response, but that improvement could potentially be added in a future version.)

In any case, the changes in this PR seem like strictly an improvement over the current behavior, which is to assume that a missing Content-Length header means there is a 0-length body, which is not a correct assumption.

propertone commented on wsgi_kerberos.py:
```diff
@@ -77,10 +91,21 @@ class KerberosAuthMiddleware(object):
     :param auth_required_callback: predicate accepting the WSGI environ
         for a request returning whether the request should be authenticated
     :type auth_required_callback: callable
+    :param read_max_on_401: When a request could not be authenticated, read and
```
This technically applies to 401 and 403. What do you think of `read_max_on_auth_fail`?

twosigmajab replied:
Done in the latest revision.

propertone commented on wsgi_kerberos.py:
```diff
@@ -40,25 +40,39 @@ def _consume_request(environ):
-        expected = int(environ.get('CONTENT_LENGTH', '0'))
+
+        content_length = environ.get('CONTENT_LENGTH')
+        if content_length is None:
```
If you do leave the Exception printing, this should also cover an empty string. At least the wsgiref server converts a missing content-length header into an empty string.

twosigmajab replied:
Thanks, handled in the latest revision.

twosigmajab commented:
> > I think WSGI servers should already validate that Content-Length headers contain valid integers before we get here
>
> I wouldn't rely on that - at least the wsgiref server doesn't do any validation.

(I think that's not surprising since wsgiref is not intended to be used in production. I meant production-quality servers, e.g. Gunicorn: https://github.com/benoitc/gunicorn/blob/548d58/gunicorn/http/message.py#L146-L150)

> I'm not sure about printing the stack trace

I originally included it because now that the exception type we're catching is sufficiently broad, when we do catch an exception, it can be helpful to see more about where it came from. That said, no longer printing the traceback in the latest revision.

> It might actually be better to just print nothing at all (as before). Unless you're a wsgi_kerberos developer, I can't imagine why devs would care about seeing debug logs of clients sending invalid content-length headers.

When swallowing exceptions like this, I think it's best practice in general to log them at debug level rather than swallow them completely silently. If there is a bug in the exception handling (e.g. wrong exception types being caught), it gives users a way of finding out about it (without having to monkey patch or modify the installed package). And users shouldn't mind when a library writes debug log messages at debug level, that's what debug level is for:)

---------

Thanks for the continued careful review! How do you think it's looking now?

propertone commented on wsgi_kerberos.py:
```diff
@@ -40,25 +40,39 @@ def _consume_request(environ):
+                    break
+                remaining -= delta
+    except Exception as exc:
+        LOG.debug("_consume_request swallowed %s", exc)
```
Nit: `_consume_request suppressed error: %s`

twosigmajab replied:
👍

propertone commented on test_wsgi_kerberos.py:
```diff
@@ -120,6 +120,16 @@ def test_unauthorized(self):
+        app = TestApp(KerberosAuthMiddleware(index, read_max_on_auth_fail=1))
+        resp = app.post('/', 'unread request body', expect_errors=True)
+        self.assertEqual(resp.status_int, 401)
+        self.assertGreater(resp.request.body_file.raw.remaining, 0)
```
This test doesn't appear to work as intended - it passes regardless of `read_max_on_auth_fail` value. Not sure what the problem is, but if it's fixable, we should have 2 tests for `len(body) > read_max_on_auth_fail` and `len(body) < read_max_on_auth_fail` to catch this.

twosigmajab replied:
Yikes, good catch, thanks! Dug through the WebTest API a bit more and it looks like the best way to create a test request with a body that we can check how much has been read after submitting the request is to use `webtest.TestRequest` to explicitly create the test request with a `body` param. After submitting the request, we can then call `tell()` on the request body file "input" object to get the number of bytes read. Done in the latest revision, which also adds tests for a bunch of different read_max values that exercise all the possible cases. Another look?

propertone reviewed (changes requested):
Almost there...

twosigmajab committed:
Limit num of bytes read in _consume_request.

Fixes #11.

Also, only consume the request for naive clients that do not send the
Expect: 100-continue header.

Finally, when no Content-Length header is sent (e.g. for chunked
Transfer-Encoding), still try to consume the request (up to the limit).
Fixes #12.

propertone reviewed (approved):
Looks good!
